gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -819,6 +825,37 @@ public FSDataOutputStream append(Path hadoopPath, int bufferSize, Progressable p
     throw new IOException("The append operation is not supported.");
   }
 
+  /**
+   * Concat existing files together.
+   * @param trg the path to the target destination.
+   * @param psrcs the paths to the sources to use for the concatenation.
+   * @throws IOException IO failure
+   */
+  @Override
+  public void concat(final Path trg, final Path [] psrcs) throws IOException {
+    logger.atFine().log("GHFS.concat: %s, %d sources", trg, psrcs.length);
+
+    Preconditions.checkArgument(psrcs.length > 0, "must be at least one source");
+
+    URI trgPath = getGcsPath(trg);
+    List<URI> srcPaths = Arrays.stream(psrcs).map(this::getGcsPath).collect(Collectors.toList());
+
+    Preconditions.checkArgument(!srcPaths.contains(trgPath), "target must not be contained in sources");
+
+    List<List<URI>> partitions = Lists.partition(srcPaths, MAX_COMPOSE_OBJECTS - 1);
```

Comment:
I think that it will be better to move partitioning logic and `MAX_COMPOSE_OBJECTS` constant in the GCS FS - it could be useful for other projects that use GCSIO library directly.

Reply:
I've moved the constant to the `GoogleCloudStorage` class.

I'm not sure the partitioning logic should be moved however, since it is tied to the Hadoop filesystem semantics. In particular, the target needs to be included in the call to compose, which may not be true of other projects implementing a compose-like operation.

Reply:
Thanks! Agree, it could be not intuitive if we will re-define GCS compose semantic in GCS IO library.